velox/connectors/clp/ClpDataSource.cpp

```diff
@@ -0,0 +1,176 @@
+  if (auto query = clpTableHandle->kqlQuery(); query && !query->empty()) {
+    kqlQuery_ = *query;
+  } else {
+    kqlQuery_ = "*";
+  }
+
+  for (const auto& outputName : outputType->names()) {
+    auto columnHandle = columnHandles.find(outputName);
+    VELOX_CHECK(
+        columnHandle != columnHandles.end(),
+        "ColumnHandle not found for output name: {}",
+        outputName);
+    auto clpColumnHandle =
+        std::dynamic_pointer_cast<ClpColumnHandle>(columnHandle->second);
+    VELOX_CHECK_NOT_NULL(
+        clpColumnHandle,
+        "ColumnHandle must be an instance of ClpColumnHandle for output name: {}",
+        outputName);
+    auto columnName = clpColumnHandle->originalColumnName();
+    auto columnType = clpColumnHandle->columnType();
+    addFieldsRecursively(columnType, columnName);
+  }
+}
+
+void ClpDataSource::addFieldsRecursively(
+    const TypePtr& columnType,
+    const std::string& parentName) {
+  if (columnType->kind() == TypeKind::ROW) {
+    const auto& rowType = columnType->asRow();
+    for (uint32_t i = 0; i < rowType.size(); ++i) {
+      const auto& childType = rowType.childAt(i);
+      const auto childName = parentName + "." + rowType.nameOf(i);
+      addFieldsRecursively(childType, childName);
+    }
+  } else {
+    search_lib::ColumnType clpColumnType = search_lib::ColumnType::Unknown;
+    switch (columnType->kind()) {
+      case TypeKind::BOOLEAN:
+        clpColumnType = search_lib::ColumnType::Boolean;
+        break;
+      case TypeKind::INTEGER:
+      case TypeKind::BIGINT:
+      case TypeKind::SMALLINT:
+      case TypeKind::TINYINT:
+        clpColumnType = search_lib::ColumnType::Integer;
+        break;
+      case TypeKind::DOUBLE:
+      case TypeKind::REAL:
+        clpColumnType = search_lib::ColumnType::Float;
+        break;
+      case TypeKind::VARCHAR:
+        clpColumnType = search_lib::ColumnType::String;
+        break;
+      case TypeKind::ARRAY:
+        clpColumnType = search_lib::ColumnType::Array;
+        break;
+      default:
+        VELOX_USER_FAIL("Type not supported: {}", columnType->name());
+    }
+    fields_.emplace_back(search_lib::Field{clpColumnType, parentName});
+  }
+}
+
+void ClpDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
+  auto clpSplit = std::dynamic_pointer_cast<ClpConnectorSplit>(split);
+
+  if (storageType_ == ClpConfig::StorageType::kFS) {
+    cursor_ = std::make_unique<search_lib::ClpCursor>(
+        clp_s::InputSource::Filesystem, clpSplit->path_);
+  } else if (storageType_ == ClpConfig::StorageType::kS3) {
+    cursor_ = std::make_unique<search_lib::ClpCursor>(
+        clp_s::InputSource::Network, clpSplit->path_);
+  }
+
+  cursor_->executeQuery(kqlQuery_, fields_);
+}
+
+VectorPtr ClpDataSource::createVector(
+    const TypePtr& type,
+    size_t size,
+    const std::vector<clp_s::BaseColumnReader*>& projectedColumns,
+    const std::shared_ptr<std::vector<uint64_t>>& filteredRows,
```

Comment:
```suggestion
    const std::shared_ptr<std::vector<uint64_t>> filteredRows,
```
Because the ownership will be transfered to `ClpVectorLoader` later